### PR TITLE
DEV-2917: Let an overlay scope inherit the grid's shortcuts, and stop cell-writing shortcuts on an unrendered body

### PR DESCRIPTION
**Pull request by budnix**
### Context

The PR fixes every grid keyboard shortcut dying while the `emptyDataState` overlay is on screen, and it does it with a mechanism any plugin can reuse.

`emptyDataState.ts` names its own shortcut context, `plugin:emptyDataState`, and hands it to the focus scope it registers. `#show()` activates that scope, and the focus scope manager switches the shortcut manager to that context. The plugin never added a single shortcut to it, and there was no fallback — `handleEventWithScope()` runs the active context only, and the `global`-scope escape hatch does not apply because both this context and `grid` are `table`-scoped. So while the overlay was up, select all, undo, redo, Tab, Enter and the arrow keys were all dead.

That is DEV-53, reported with two ways to reach the state through the UI:

- Select all cells, hide every column, click where a row header is — Cmd/Ctrl+A no longer selects anything.
- Select all rows, remove them — Cmd/Ctrl+Z no longer undoes, although Undo from the context menu does.

**The first version of this PR listed the missing keys in the plugin.** That was rejected in review for the right reason: a shortcut added to the grid next year would be silently dead in the overlay until someone remembered to edit this one file. This version replaces the list with inheritance, and the four commits below are one layer each.

### What changed

**1. A focus scope can declare a fallback shortcut context** (`FocusScopeOptions.fallbackShortcutsContextName`). The shortcut manager walks that chain when it dispatches a key, so a scope that *covers* the grid without replacing it answers everything the grid answers — including shortcuts added later, with no change to the plugin. A **modal** scope (`dialog`, the license lock screen) leaves it unset, because letting grid shortcuts through a modal is the bug, not the feature.

**The walk moves on only when nothing in the current context ran.** A shortcut whose `runOnlyIf` returned `true` stops it, whether or not its callback did anything; a shortcut that is registered but declines does not shadow the fallback's answer for the same key. Without that rule the plugin's own `Ctrl`+`A` shadowed the grid's for good — a context could kill a key by declaring it and then refusing it, which is exactly the rot inheritance is meant to avoid. The rule is written into the manager's JSDoc, because the two readings are easy to confuse.

The chain is walked by the manager, deliberately not by the context: `hasShortcut()` and `getShortcuts()` stay own-only, because they are public API documented as answering for one context and `ShortcutManager#hasEventShortcut()` (which `comments` calls) depends on that. The walk is cycle-guarded, and its `Set` is allocated only once a fallback is really walked — this runs on every keydown. `runGlobalScopedShortcuts` does not walk at all: it runs while the table pipeline is blocked, so a global context with a fallback could otherwise have reached the very table-scoped shortcuts that block exists to stop.

`emptyDataState` opts in with one line and registers exactly one shortcut of its own — select all, as an **override**, because the grid's own `Ctrl`+`A` is guarded on `isDefined(getSelected())` and the overlay starts with nothing selected, which is precisely the case that matters. It reuses the grid's own `selectAllCells` command rather than repeating it.

**2. A shortcut that reads or writes cell content now calls `canAccessCellContent(hot)`** (`shortcuts/guards.ts`).

This half is what makes the first half safe, and it is a real bug of its own. Measured on a grid with every column hidden: `Ctrl`+`A` then **Delete blanked all 40 cells**, and **Ctrl+Enter overwrote all 40** with the highlight's value. Ctrl+Enter got there for a reason worth knowing — `Context#addShortcuts(shortcuts, options)` copies a group option onto a shortcut only when the shortcut does not already declare that key, so a per-shortcut `runOnlyIf` **replaces** the group's guard instead of being ANDed with it. The `GRID_GROUP` rendered-cells check reads like a floor and is not one.

The guard now lives in its own module rather than in a closure inside `contexts/grid.ts`, because **two other files register destructive shortcuts into that same grid context and could not reach it**: `mergeCells` (`Ctrl`+`M`, which clears every cell but the top-left one when un-merging) and `checkboxRenderer` (space, enter, delete, backspace). Both take it now, each writing it into its own `runOnlyIf` for the reason above.

**The guard asks two questions, because one is not enough.** *Are cells drawn* — and *is anything covering them*. An earlier revision of this PR asked only the first, and that was a data-loss bug in its own right: `emptyDataState` does not only show its overlay on an empty grid. The `#loadingActive` branch of `#toggleEmptyDataState()` shows it over a **fully rendered** grid while a DataProvider fetch runs — the method's own comment says so. There the drawing check answered `yes`, so the shortcuts this PR makes the overlay inherit reached the data underneath. Measured: 8 rows and 6 columns still drawn, selection alive, and **Delete wiped the dataset**; removing the `fallbackShortcutsContextName` line alone made it inert again, so it was ours.

A focus scope therefore declares `coversGridBody`, and the manager answers `isGridBodyCovered()` from every **enabled** scope rather than the active one — an overlay is on screen whether or not it holds the keyboard, and that is what puts the cells out of reach.

Two write paths are outside this guard by construction, and no shortcut guard can reach them: `copyPaste` binds native `paste`/`cut` DOM listeners, and anything the host calls through the API. Both are stated in the guide and in `AGENTS.md`. Pasting under the loading overlay does write — measured — but it does so identically with and without this PR, so it is pre-existing and belongs on its own ticket. One gap is deliberately left: the license lock screen declares neither `coversGridBody` nor an activated scope, so keyboard editing still works behind the lock. Pre-existing, named in `handsontable/AGENTS.md`, and not something to change inside a keyboard-shortcut PR.

Shortcuts that only *move* the selection deliberately do not take the guard — navigating the headers still on screen is useful and destroys nothing. The `Core` methods keep guarding on the data counts, so `emptySelectedCells()` called through the API still clears hidden columns; only the keystroke is gated.

**3. Tab no longer dies inside an overlay.** The `GRID_TAB_NAVIGATION_GROUP` pair ran with **no guard at all**, and its `after()` calls `event.preventDefault()` whenever the selection is still in range. So an overlay that inherits the grid's shortcuts swallowed Tab, and the user had no way out of it — measured: `defaultPrevented: true`, focus stuck.

**The fix is not a guard on that pair, and an earlier revision of this PR got that wrong.** The pair is bookkeeping: `before()` sets state that `after()` clears, and every guard worth having here can change its answer *during* the keystroke, because the Tab move itself drops the selection. A `runOnlyIf` there runs one half and skips the other, leaving `tabNavigation`'s `isTabOrShiftTabPressed` set so its `afterSelection` hook forces `preventScrolling` on a later, unrelated selection. Both pairs — this one and the identical second pair in `focusManager/scopes/grid.ts`, which a review rightly pointed out was never guarded — are now left alone, and `!canNavigateGrid(hot)` joins the existing deselect branch inside `after()`, which is where the `preventDefault()` that traps the user actually lives.

`canNavigateGrid()` shares the covering half with the cell-content guard and is weaker on the other half: `navigableHeaders` or drawn cells, because moving a selection into a visible header is legitimate where writing to a hidden cell is not. Both `navigableHeaders` settings and the loading overlay are pinned by their own tests.

**4. Three more scope-manager fixes, all from review.** `activateScope()` captured the displaced context *after* `scope.activate()` — and `onActivate` commonly moves DOM focus, which re-enters `processScopes()`, so a deactivation in that window restored nothing and stranded the manager on the plugin's context with no active scope. `registerScope()` now throws when a scope falls back to its own context, which is what `fallbackShortcutsContextName: 'grid'` with no `shortcutsContextName` produced, since that option defaults to `'grid'`. And `unregisterScope()` clears the fallback only when no other scope still declares the same pair, because the fallback lives on the shared context. Separately, `emptyDataState.updatePlugin()` re-activates its scope only when that scope was already active — it used to take the keyboard from an open modal dialog on any `updateSettings`.

**5. The shortcut context rolls back in the scope manager, not in the plugin.** A scope now records the context it displaced when it was activated, and `deactivateScope()` restores it — held per scope so nesting unwinds in order, and skipped when something else (an open editor) took the context over meanwhile. `#hide()` keeps no rollback of its own; the previous version hardcoded `'grid'`, which clobbers an open editor's context, and `disablePlugin()` could not reach it at all, so `updateSettings({ emptyDataState: false })` recreated the original bug. `unregisterScope()` now deactivates an active scope before destroying it, and `updatePlugin()` re-activates the scope when the overlay is still on screen.

Every deactivation restores, **including** the one where focus simply left every scope. Skipping that path looks harmless — the grid is unlistened there, so the name decides nothing — but the stale name survives and the *next* activation records it as the context it displaced. The scope then rolls back to its own name and the grid is stuck exactly as before. I tried the narrow version first and this is what it does; there is now a test that pins it.

**One behavior change in the legacy suite.** 40 assertions across the `dialog`, `emptyDataState` and `pagination` Tab specs read `'plugin:X'` at points where focus had already left the grid. They now read `'grid'`. They pinned dead state: the grid does not listen while focus is outside it, so no shortcut from either context could run there. One `scopeManager` assertion pinned the defect itself — the context staying on the plugin's name after an explicit `deactivateScope()` — and now expects `'grid'`.

Not a breaking change. The plugin context was empty, so there is no behavior to preserve; the grid guards only block keystrokes that acted on cells nobody could see, and Tab now does what the browser would have done.

`pagination` has the same empty-context gap and can now opt in with one line. Out of scope here.

One visible behavior change worth calling out: undo and redo now work **during** a DataProvider fetch, because the overlay inherits them. That is what DEV-53 asked for, and it is a deliberate consequence rather than an oversight — cell-content shortcuts stay blocked in that state, so the data under the overlay cannot be edited, only the history moved.

Two further findings from DEV-53 stay on that task: with every column hidden there is no way back through the UI at all (no context menu over the overlay, no headers, and hiding columns is not undoable — needs a product decision), and the selection restored after undoing a full row removal is out of range. One more was seen while writing the checkbox test: undo after toggling a mixed checkbox selection leaves the last row flipped. Not touched here.

### Test evidence (required for source changes)

- Unit tests added/modified (`*.unit.js`): `handsontable/src/shortcuts/__tests__/manager.unit.ts` — 12 tests, including the fallback resolving, an own entry winning, **falling through when the own entry declines**, a shortcut that ran stopping the walk even when its callback did nothing, a chain of fallbacks, a cycle visiting each context at most once, `hasShortcut()` still answering for one context, and a global context not reaching a table-scoped fallback
- E2E tests added/modified (Playwright `tests/e2e/*.spec.ts`): `tests/e2e/empty-data-state-shortcuts.spec.ts` — 15 tests, with `tests/fixtures/demo/empty-data-state-shortcuts.html` and `tests/fixtures/pages/EmptyDataStateShortcutsPage.ts`
- Legacy E2E modified: `handsontable/src/focusManager/__tests__/scopeManager.spec.js` (one assertion corrected, one new focus-round-trip test), plus the 40 assertions described above
- Type tests (`*.types.ts`) updated if public API changed: `handsontable/src/focusManager/__tests__/scopeManager.types.ts` — `fallbackShortcutsContextName`
- For a bug fix — the spec that fails without this fix: **each layer was reverted alone, and each control is exact.** Reverting `shortcuts/guards.ts` alone fails exactly the four data-protection tests (Delete, Ctrl+Enter, Ctrl+M, Space) and nothing else. Reverting `coversGridBody: true` alone fails exactly the two loading-overlay tests and leaves those four green. Putting the displaced-context capture back after `activate()` fails exactly the stranding test, and re-activating the scope unconditionally fails exactly the dialog test. Reverting the Tab guard alone fails exactly the Tab test. Stopping the walk at the first context that holds the key fails exactly the two fall-through unit tests. Disabling the context rollback fails exactly the two `scopeManager` specs and "restores the grid shortcut context once the overlay hides".
- Demo page / recorded trace (for UI changes): none

Every negative test carries a positive control in the same test, so none of them can pass on a grid where the key never worked at all, and each settles on a real state change (the selection clearing) rather than on a timer. One test is a plain regression pin rather than a control, and says so in its own comment: the `navigableHeaders: true` Tab case passes with the guard reverted too, because the pair ran unguarded there as well.

### Commands run

```bash
npm --prefix handsontable run test:unit
```
`Test Suites: 377 passed, 377 total` / `Tests: 5101 passed, 5101 total`

```bash
npm --prefix handsontable run test:e2e
```
`9966 specs, 0 failures, 8 pending specs` (the full legacy suite)

```bash
npm --prefix handsontable run test:walkontable
```
`845 specs, 0 failures`

```bash
npm --prefix tests run test:e2e -- e2e/empty-data-state-shortcuts.spec.ts
```
`90 passed` — all six theme × bundle legs, on a full `npm --prefix handsontable run build`

```bash
npm --prefix handsontable run test:types
npm --prefix handsontable run eslint
npm --prefix handsontable run stylelint
npm --prefix tests run lint
```
types ✓ / `1924 problems (0 errors, 1924 warnings)` — every warning pre-existing and in unrelated files / stylelint ✓ / tests package clean

One flake was seen once and did not reproduce in two further full runs: `Selection running in iframe should set correct wtBorder top position of selected cell` (`Expected '1px' to equal '29px'`), which the runner itself recorded as "passes alone". It waits on `waitForNextAnimationFrames(2)` — a fixed frame count, the shape the repo's own lint rule warns about — and it passes on `develop` and on this branch. The same test flaked once in the previous round.

Two Jasmine cases were added to the existing `scopeManager.spec.js` rather than to Playwright. Editing a frozen spec is allowed, and these drive that file's own `createUIWithFocusScope` helper and assert focus-manager internals no page can observe. The dialog pair went to Playwright — the pre-push gate rejected a new `*.spec.js`, correctly.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):

1. DEV-2917
2. DEV-53
3.

### Affected project(s):

- [x] `handsontable`
- [ ] `@handsontable/angular-wrapper`
- [ ] `@handsontable/react-wrapper`
- [ ] `@handsontable/vue3`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://cla.handsontable.com/sign) — one signature covers both Handsontable and HyperFormula; the `cla/signed` check on this PR confirms it.
- [x] My change requires a change to the documentation.
- [ ] MANUAL QA NEEDED — 

ClickUp task: https://app.clickup.com/t/9015210959/DEV-2917

---

> [!NOTE]
> **Medium Risk**
> Touches focus-scope activation, shortcut context switching, and grid-wide keyboard guards—high user-impact paths with broad test coverage but subtle interaction edge cases.
> 
> **Overview**
> Fixes grid keyboard shortcuts dying while the **empty data state** overlay (or a DataProvider loading overlay) is visible, without hard-coding shortcut lists in the plugin.
> 
> **Focus scopes** gain `fallbackShortcutsContextName` (e.g. `'grid'`) and `coversGridBody`. The shortcut manager walks the fallback chain when the active context has no handler; explicit `deactivateScope` / `unregisterScope` restores the displaced shortcuts context, while focus-driven deactivation does not.
> 
> **Guards** in `shortcuts/guards.ts` gate cell-writing shortcuts with `canAccessCellContent()` (rendered cells + body not covered) and navigation with `canNavigateGrid()` plus `keepCoveredCellsUnselectable()` so selection cannot land on covered cells. Tab’s `preventDefault` is gated so users can Tab out of overlays.
> 
> **emptyDataState** inherits grid shortcuts, sets `coversGridBody: true`, and overrides **Ctrl/Cmd+A** only when nothing is rendered. **mergeCells** and **checkboxRenderer** shortcuts use the same content guard.
> 
> Docs, plugin dev guidance, changelog **#13489**, unit/E2E coverage (including a new Playwright suite).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7bbc3bab66a475a6896fd4bd81174d63fc2ec924. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qunabu** [comment]
Task linked: [DEV-2917 EmptyDataState swallows every keyboard shortcut while the overlay is shown](https://app.clickup.com/t/9015210959/DEV-2917)
Task linked: [DEV-53 Keyboard shortcuts not working after actions in emptyState plugin](https://app.clickup.com/t/9015210959/DEV-53)

File: handsontable/src/focusManager/scopeManager.ts [hunk 1]
```diff
@@ -138,7 +163,37 @@ export function createFocusScopeManager(hotInstance: HotInstance): FocusScopeMan
...
+    if (fallbackContextName === contextName) {
+      SCOPES.removeItem(scopeId);
+      scope.destroy();
+      throwWithCause(`The "${scopeId}" focus scope falls back to its own shortcuts context ` +
+        `("${contextName}"). \`shortcutsContextName\` defaults to "grid", so a scope that declares only ` +
+        '`fallbackShortcutsContextName` names the same context twice.');
+    }
+
+    if (fallbackContextName !== null) {
+      scopeContext.setFallbackContext(shortcutManager.getOrCreateContext(fallbackContextName));
```
Comment (Haxikowy) [inline handsontable/src/focusManager/scopeManager.ts:197]: If two scopes share the same `shortcutsContextName` but declare different `fallbackShortcutsContextName`s, this silently overwrites the first scope's fallback pointer with the second's — no error, unlike the self-loop check just above. `unregisterScope`'s `hasOtherScopeWithSameFallback` (line 192) then compares by name equality, so tearing down the first scope can null out the fallback the second scope still needs.

File: .changelogs/13489.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+  "issuesOrigin": "private",
+  "title": "Fixed keyboard shortcuts, including select all, undo and redo, not working while the empty data state overlay is on screen. Shortcuts that change cell content no longer run while the grid renders no cells or an overlay covers them. Tab now moves the focus out of the overlay.",
```
Comment (Haxikowy) [inline .changelogs/13489.json]: This title doesn't mention the loading/DataProvider-fetch case — the one behavior change the PR body itself calls out as user-visible (undo/redo now work during a fetch). A customer debugging that behavior would search the changelog for "loading" or "data fetch," not "empty data state."

**Review by Haxikowy — COMMENTED** [review]
Off-diff findings (outside any changed hunk, or no clean single-line anchor):

**`Ctrl`+`A` never got the new guard** (`handsontable/src/shortcuts/contexts/grid.ts`, lines 57-63, unchanged by this diff). Every other shortcut in this file that carries its own `runOnlyIf` was updated to call `canAccessCells()`/`canNavigateGrid()` — `Ctrl`+`Enter`, `Alt`+`Enter`, the arrow+`Shift`+`Ctrl` combos, `Home`, `End`. `Ctrl`+`A`'s two entries still only check `isDefined(hot.getSelected())` and header status. Under `emptyDataState`'s `coversGridBody` overlay, with a selection still alive, `Ctrl`+`A` still runs `selectAllCells()` on cells the user can't see — the same class of bug this PR just fixed for `Ctrl`+`Enter`.

**Plain arrow keys have no post-move check** (`handsontable/src/shortcuts/contexts/grid.ts`, lines 98-99 `ArrowUp`/`ArrowDown` and their `Left`/`Right` siblings, unchanged by this diff). Tab re-checks `canNavigateGrid()` in `after()` once the move lands and deselects if it's bad. Plain arrow keys only gate entry through the group's `isGridNavigable()` and never re-check the destination. With `navigableHeaders: true` and a covering overlay, moving down from a header selection could land on a covered, invisible cell with nothing to clear it. Flagged from static reading only — worth a quick repro before treating as blocking.

**Test-evidence counts don't match this diff.** The PR body's headline breaking-change evidence — "40 assertions across the dialog, emptyDataState and pagination Tab specs" changed from `'plugin:X'` to `'grid'` — isn't in any of the 26 changed files. `scopeManager.spec.js` gets 7 new `it()` blocks (claimed: 1), `manager.unit.ts` gets 8-9 new tests (claimed: 12), and the new E2E spec has 17 tests (claimed: 15). Worth reconciling before merge, since reviewers can't verify the 40-assertion claim from this branch.

**Breaking-change note:** `deactivateScope`/`unregisterScope` on `FocusScopeManager` (public API via `hot.getFocusScopeManager()`) used to leave the active shortcuts context untouched; now they actively roll it back. The PR's breaking-change section only reasons about the new `fallbackShortcutsContextName` option, not this behavior change on an existing method. Worth a line in the description per the repo's breaking-changes policy, even if the conclusion stays "not breaking."

File: handsontable/src/focusManager/scopeManager.ts (hunk 1 above)
Comment (budnix) [reply handsontable/src/focusManager/scopeManager.ts:197]: Good spot — `registerScope` now refuses that pair in 187ff86aee, the same way the self-referencing fallback just above it is refused.

File: .changelogs/13489.json (hunk 2 above)
Comment (budnix) [reply .changelogs/13489.json]: Fair point — the title now names the data provider fetch too, in 187ff86aee.

**Review by Haxikowy — APPROVED** [review]
